internal/stats/query.go

```diff
@@ -31,14 +27,72 @@ type BaseQuery struct {
 	// Define how the results should be sorted by time.
 	// -1 for ascending, 0 for no sorting, 1 for descending.
 	Sort int `json:"sort"`
+
+	// Filter results to only include metrics that match Metric attribute field values.
+	// Keys are field names (e.g., "course") and values are what to include (e.g., course101).
+	// This filter is applied after all other Query conditions are applied.
+	Where map[MetricAttribute]any `json:"where,omitempty"`
+
+	// Only return data of this type.
+	// This field is required in the query to specify which kind of metric to return.
+	Type MetricType `json:"type"`
+}
+
+func (this *Query) Validate() error {
+	if this == nil {
+		return fmt.Errorf("No query was given.")
+	}
+
+	if this.Type == Unknown_Metric_Attribute_Type {
+		return fmt.Errorf("Query type was not set.")
+	}
+
+	if this.Where == nil {
+		this.Where = make(map[MetricAttribute]any)
+	}
+
+	for field, value := range this.Where {
+		if field == Unknown_Metric_Attribute_Key {
+			return fmt.Errorf("Query attribute field was empty.")
+		}
+
+		if value == nil || value == "" {
+			return fmt.Errorf("Query attribute value was empty.")
+		}
+	}
```

Comment:
The Type and Where checks look the same between metrics and queries.
I would make a common validation function for these.